documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,366 @@
+
+# I types versus F Types
+Each type in this proposal, aside from `Quaternion` ends in I or F, defining whether it is an integer type or floating point type. Integer types **must** use a generic type argument `T` with the constraint of `IBinaryInteger<T>`. On the other hand, floating point types **must** use a generic type argument `T` with the constraint of `IFloatingPointIeee754<T>`.
+
+# Vector Types
+
+The main types defined for this proposal are two sets of vector types, `VectorNI<T>` and `VectorNF<T>` where `N` defines the dimensionality of the vector between 2 and 4.
+
+For each vector struct, the following requirements **must** fulfill the following requirements:
+- Implements IEquatable with itself as the generic parameter
+- Implements IReadonlyList with the components as the list elements
+- The relevant number of properties to represent the mathematical vector's components (X and Y for Vector2) and relevant unit vectors
+- Constructors which take either a single parameter and uses it for every component, a parameter for each component, or a ReadOnlySpan of values which has the same number of elements as our vector has components.
+- Constructors for 3 dimensions and up **must** include lower dimension variants that use the lower dimensions for their specific components (vector2 -> X,Y).
+- A ref indexer that takes a int index and returns the corresponding component value (0 -> x, 1 -> y, etc.)
+- An AsSpan function which returns this vector as a Span of the generic type
+- A LengthSquared property which returns the dot product of the vector with itself.
+- A Dot function which takes another vector and returns the dot product with our original vector. 
+  - A static implementation of this function **must** be available as well.
+- For 3D Vectors, a Cross function which takes another vector and returns the cross product with our original vector. 
+  - A static implementation of this function **must** be available as well.
+- `+`, `-`, `*`, `/`, and `%` operators defined between two vectors of the same type which returns a vector which has had each operation applied component-wise.
+- `+`, `-`, `*`, `/`, and `%` operators defined between a vector and a scalar value that matches the generic type which returns a vector which has had each operation applied component-wise with the scalar value. Both vector first and scalar first should be implemented.
+- A `-` unary operator which returns the negated vector.
+- A `+` unary operator which returns the vector.
+- Overrides ToString to show component values.
+- Include additional ToString implementations which take a `string` format and a `string` format and `IFormatProvider`
+- TryFormat functions which output a formatted string of the vector to a `Span<char>` or `Span<byte>`, in UTF-16 and UTF-8, respectively.
+- Static Parse and TryParse functions which take a `string`, `Span<char>`, or `Span<byte>` and `IFormatProvider`.
```

Comment:
notate the implementation of ISpanFormattable, ISpanParsable, and IUtf8... interfaces?

Reply:
Good catch, changing to just implementing those interfaces now